File: 완전탐색/program_다단계칫솔.py
def game(cash, user, sell, cost):
    orzi = cost//10
    rest = cost - cost//10
    if cost>=10:
        cash[sell] += rest
        if user[sell]:
            game(cash, user, user[sell], orzi)
    else:
        cash[sell] += cost
    

def solution(enroll, referral, seller, amount):
    user = {
        "minho":[]
    }
    cash = {
        "minho":0
    }
    for i, name in enumerate(enroll):
        if name not in user:
            user[name] = []
            cash[name] = 0
        if referral[i] == "-":
            user[name] = "minho"
        else:
            user[name] = referral[i]
    for i, sell in enumerate(seller):
        game(cash, user, sell, amount[i]*100)
    
    answer = []
    for name in enroll:
        answer.append(cash[name])
    return answer

File: 완전탐색/program_다단계칫솔.java
import java.util.*;
import java.io.*;


public class Solution {

    public void game(HashMap<String, String> user, HashMap<String, Integer> cash, String sell, Integer cost){
        Integer nextCost = (Integer) cost/10;
        Integer remain = cost-nextCost;
        if(cost>=10){
            cash.put(sell, cash.get(sell) + remain);
            if(user.get(sell) != null)
                game(user, cash, user.get(sell), nextCost);
        }
        else{
            cash.put(sell, cash.get(sell) + cost);
        }

    }

    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        HashMap<String, String> user = new HashMap();
        HashMap<String, Integer> cash = new HashMap();

        for(int i=0; i< enroll.length; i++){
            if(referral[i].equals("-"))
                user.put(enroll[i],null);
            else
                user.put(enroll[i],referral[i]);
            cash.put(enroll[i],0);
        }

        for(int i=0; i< seller.length; i++){
            game(user, cash, seller[i], amount[i]*100);
        }

        int[] answer = new int [enroll.length];
        for(int i=0; i<enroll.length; i++){
            answer[i] = cash.get(enroll[i]);
        }
        System.out.println(Arrays.toString(answer));
        return answer;
    }


    public static void main(String[] args) throws Exception {
        Solution sol = new Solution();
        String[] a = {"john", "mary", "edward", "sam", "emily", "jaimie", "tod", "young"};
        String[] b = {"-", "-", "mary", "edward", "mary", "mary", "jaimie", "edward"};
        String[] c = {"young", "john", "tod", "emily", "mary"};
        int[] d = {12, 4, 2, 5, 10};
        System.out.println(sol.solution(a,b,c,d));
    }
}
